gui/default/syncthing/core/syncthingController.js

```diff
@@ -584,6 +626,19 @@ angular.module('syncthing.core')
             }).error($scope.emitHTTPError);
         }
 
+        $scope.probeRemoteGUIAddress = function (deviceId, address) {
+            $http({
+                method: "OPTIONS",
+                url: address,
+            }).success(function (data) {
+                $scope.remoteGUICache[address] = true;
+                $scope.idToRemoteGUI[deviceId] = address;
+            }).error(function (err) {
+                $scope.remoteGUICache[address] = false;
+                $scope.idToRemoteGUI[deviceId] = "";
```

Comment:
Originally it was meant to cache for a longer duration than `idToRemoteGUI()`, to avoid re-probing the same address frequently.  However, because we clean the cache during each `refreshConnectionStats()` call, there's really no gain I suppose, and it's just leftover from @AudriusButkevicius' pseudocode suggestion.

I think re-probing is the right thing to do actually, even if the protocol address changes to something else and back to a previous value.  Something would have triggered that connection change, and the same reason might also affect the GUI reachability.

Reply:
Re-thinking that, there is still the case when the connection was previously probed with error, so the `idToRemoteGUI` object does not include it and we would re-probe over and over, because any valid address is different from `""`.

Reply:
However `idToRemoteGUI` is assigned too when there's a probing error. The way I read the code the probing is triggered only when coming online and when refreshing the window. It does not re-probe addresses regardless of whether probing succeeded or not, unless the address vanishes and reappears. And the bool value of `remoteGUICache` is only assigned, but never used.

Reply:
> The way I read the code the probing is triggered only when coming online and when refreshing the window.

`refresh()` is called every 10 seconds, which in turn triggers `refreshConnectionStats()`.

> It does not re-probe addresses regardless of whether probing succeeded or not, unless the address vanishes and reappears.

Yes, we assume that probing once is enough per address, until that address becomes stale and we discard the previous probing result. 

> And the bool value of `remoteGUICache` is only assigned, but never used.

You're right the value does not matter, but some value is needed for the property in an Object, so what better to use than the result status? (Good for debugging.)  The mere presence of the key in `remoteGUICache` is what matters.

`idToRemoteGUI` is set to the empty string on probing error. So without the cache, we wouldn't know the already probed address.  Do you have an idea how to unify those two structures?

Reply:
Thanks for the explanations, makes sense now.

Reply:
I just tested a simpler idea that gets away without manipulating the cache from the HTTP request callbacks.  Saves the cleanup loop at the expense of the true / false result indication.

@rjpruitt16 feel free to look at [my branch](https://github.com/acolomb/syncthing/tree/gui-add-advance-config-port-mapping).  You could also give me write access on your fork in order to contribute changes directly to this PR if you like.

Reply:
@acolomb just sent collaboration request